FIX #15739 dynamic price

atm-florianm opened the pull request:
# Fix #15739 
A dynamic (selling) price is computed using an expression that can rely on several values, including minimum selling price, best buying price, weight, etc.

When the `Product` object is fetched, the price is automatically recomputed, which explains why the price displayed on a product card is always up to date. However, when it is recomputed, it is not always stored, so pages that rely on SQL queries to get a product's selling price (such as the product list) show a potentially outdated price value.

The price of a product using a dynamic price should be updated (in the database) whenever one of the values used its price expression (or the expression itself) changes.

This PR is not complete in that respect, but it lays the foundation for further fixes.

atm-florianm committed:
FIX: do not allow empty mandatory values when editing global variables for dynamic prices

atm-florianm committed:
FIX 12.0 (dynamic prices): help text shows wrong substitution key for variable (it shows "globals_xyz" instead of "global_xyz")

atm-florianm committed:
FIX 12.0 dynamic prices: when updating a component used in dynamic price calculation, recompute the price and save it into db.

atm-florianm committed:
Merge branch '12.0' into fix/12.0_issue_15739_dynamic_price

atm-florianm committed:
Update product.class.php

phpdoc

eldy commented on htdocs/fourn/class/fournisseur.product.class.php:
```diff
@@ -383,6 +383,12 @@ public function update_buyprice($qty, $buyprice, $user, $price_base_type, $fourn
+						// Computed selling prices may depend on the purchase price. By re-fetching the product, we
+						// force it to re-compute the selling price and to save it if it has changed.
+						$product = new Product($this->db);
+						if ($product->fetch($this->id) > 0) $product->updatePriceFromExpression();
```
Can you confirm you need to create a new object Product.
Why not just $this->updatePriceFromExpression()

atm-florianm replied:
Sorry, this was the remnant of an older code layout which required it, but it is no longer necessary. I modified the code accordingly.

eldy commented on htdocs/product/price.php:
```diff
@@ -378,6 +378,7 @@
 			$action = '';
 			setEventMessages($langs->trans("RecordSaved"), null, 'mesgs');
 			$db->commit();
+			$object->updatePriceFromExpression();
```
This is outside the transaction ?  After the commit ? Why ?

atm-florianm replied:
OK

atm-florianm committed:
FIX PR#15909:
+ error handling
+ updatePriceFromExpression before DB commit
+ remove useless product instantiation

atm-florianm committed:
Merge branch 'fix/12.0_issue_15739_dynamic_price' of github.com:atm-florianm/dolibarr into fix/12.0_issue_15739_dynamic_price

eldy commented:
Because this is an experimental feature, we have no reason to fix it in v12. Becasue, it seems there is some changes related to same part of code into v13, this will break v13 too. Can you try to push your PR to v13 or develop branch ? (by merging the conflict due to other changes on same files) ?

eldy commented:
This PR is now too old, so i close it. A lot of code may have been validated since the PR and the PR may be obsolete, or too difficult to merge due to other changes. Best choice is too cancel it to avoid to keep PR with too low chances to be validated.
